Mup 1.5

zodern opened the pull request:
Mup 1.5 adds load balancing, zero downtime deploys, performance improvements, and can optionally use Docker Swarm.

### Current Status

**Docker Swarm** Swarm is used for load balancing and zero downtime deploys. Mup is able to setup and manage the swarm cluster. There are a couple situations it doesn't correctly handle yet, listed in #893.

**Docker Services** When swarm is enabled, the app is deployed as a swarm service. The basics work, but there is a lot left to do. Many commands need to be updated to work with swarm services, and the commands that have been might be missing features or not work correctly.

**Load Balancing** Uses [nginx-proxy-swarm-upstream](https://github.com/zodern/nginx-proxy-swarm-upstream). The current implementation works but isn't the best way to get docker-gen to support swarm. In the future I will try to add swarm support to docker-gen.

**Zero downtime deploys** In progress

**Changing default docker image** The current default docker image only supports Meteor 1.3 and older, so most users need to modify the config to use the correct docker image. I have been working on the [`zodern/meteor`](https://github.com/zodern/meteor-docker) docker image which supports Meteor 1.2 and newer. Once it is compatible with the parts of meteord that Meteor Up uses, `zodern/meteor:root` will become the default image.

**Performance** Prepare Bundle will be much faster for apps with a large number of environment variables, and the Start Meteor task list will finish 12 seconds sooner when Prepare Bundle is enabled. The `Mongo Start` task list will many times only be a couple of seconds instead of around 30. Also, ssh sessions are kept connected between task lists instead of disconnecting and reconnecting. Mup's tests finish over 15 minutes faster with mup 1.5.

#### Beta

You can try the beta with `npm i -g mup@next`. You can view the new docs [here](https://github.com/zodern/meteor-up/pull/975/files?short_path=c331e99#diff-c331e99b809715805fe7447af020d198), and the list of changes in the [changelog](https://github.com/zodern/meteor-up/pull/975/files?short_path=c331e99#diff-4ac32a78649ca5bdd8e0ba38b7006a1e).

MasterJames commented:
Maybe it's still relevant to submit to the current port at kadirahq/meteor-up?  So to say this version is depreciated and the "Stable" version, is to say it's not changed or merged PRs for a long time now, so please consider adding to a branch over there.

vhf commented:
Makes sense. I'll look into kadirahq/meteor-up at work next week. If it's still not configurable, I'll see what I can do to submit a patch there.

zodern committed:
Merge branch 'master' into dev

zodern committed:
Merge branch 'dev' into origin/mup-1.5

zodern committed:
Run meteor app as swarm service

zodern committed:
Run meteor app as swarm service

zodern committed:
Simplify swarm api's

- Use server's name instead of host in more places
- Labels returned by swarmOptions is an array instead of an object

zodern committed:
Simplify swarm api's

- Use server's name instead of host in more places
- Labels returned by swarmOptions is an array instead of an object

zodern committed:
Store validation errors even when they aren't shown

zodern committed:
Add noDefaults to VALIDATE_OPTIONS

zodern committed:
Fix starting app

zodern committed:
Update eslint config to warn if no new line after var

zodern committed:
Add api.tasks with tasks for docker swarm

zodern committed:
Allow using nodemiral session in runSSHCommand

zodern committed:
Add pluginApi.dockerServiceInfo

zodern committed:
Use addCreateOrUpdateService for starting the Meteor app

zodern committed:
Increase concurrency of serverInfo to 3

zodern committed:
Update eslint config

zodern committed:
Merge remote-tracking branch 'origin/mup-1.5' into mup-1.5

zodern committed:
Re-use sessions between task lists and runSSHCommand

zodern committed:
Add stop service task

zodern committed:
Add load balancing

zodern committed:
Disable checking for updates in tests

zodern committed:
Fix running tasks on managers when setting up swarm

zodern committed:
Update Meteor app used in tests to 1.7
A number of packages were also removed to help speed up tests.

zodern committed:
Fix proxy status when proxy.servers in the config is used

zodern committed:
Show better message when unable to add or remove a node label

zodern committed:
Fix some managers not being promoted after initializing swarm

zodern committed:
Remove unused props and exports in swarm utils

zodern committed:
Throw error when servers are in multiple clusters
Any servers already in a cluster need to be in the same cluster. This tries to show helpful information so the user can decide how to fix it.

I had considered having mup try to automatically fix this, but that could easily cause other clusters or apps to stop working.

zodern committed:
Improve task name when destroying swarm cluster

zodern committed:
docker status shows "Unknown server" for nodes not in the config

zodern committed:
Safely demote nodes

zodern committed:
Use smaller image in tests to fix travis no space on disk error

zodern committed:
Node being joined should not already be part of cluster

zodern committed:
Validate swarm object in config

zodern committed:
Document load balancing and swarm

zodern committed:
Add 'nan' to list of packages to remove from shrinkwrap

zodern committed:
Fix lint error

zodern committed:
Document minimum memory servers should have when using Swarm

zodern committed:
Fix syslog test

zodern committed:
app.docker.imagePort defaults to 3000 instead of 80All of the docker images I know of that are compatible with Meteor Up allow using the PORT environment variable to change what port the app uses, so this change should not break compatibility.

zodern committed:
Fix docker ps
Since sessions are now closed after the last promised is resolved, commands and hooks using sessions need to return a promise.

zodern committed:
Remove unused swarmServices from server info

zodern committed:
Remove unused task

zodern committed:
Wait less time after starting the app and prepare bundle is enabled

zodern committed:
Only restart mongodb container when necessary
The container is started when it isn't running, or the start script has changed.

zodern committed:
Set NODE_VERSION build arg during prepare bundle

zodern committed:
Combine ENV build instructions in Prepare Bundle
This saved ~30 seconds in one of mup's tests

zodern committed:
Fix verifying deployment with non-root user

zodern committed:
Use title case for stop mongo task name

zodern committed:
Update changelog

zodern committed:
Revert "Update Meteor app used in tests to 1.7 A number of packages were also removed to help speed up tests."

This reverts commit 27b81807795b035d9c62fb8f47472fea14d2fd51.
The Meteor 1.7 app takes over 90 seconds longer to build for the tests.

zodern committed:
Remove newline after node version from .node_version.txt

zodern committed:
Fix load balancing services listening on ports other than 80

zodern committed:
Update changelog

zodern committed:
Merge branch 'master' into mup-1.5

zodern commented:
I created #976 with the correct branch.
